_books/ion-1-1/src/macros/macros_by_example.md

```diff
@@ -155,720 +132,160 @@ The outer e-expression receives the results from the expansion of the inner e-ex
-  ⇒ (:make_string a (:values b (:values c) d) e)
-//                               ^^^^^^ next
-  ⇒ (:make_string a (:values b c d) e)
-//                    ^^^^^^ next
-  ⇒ (:make_string a b c d e)
-  ⇒ "abcde"
-```
-
-Splicing within sequences is straightforward, but structs are trickier due to their key/value
-nature. When used in field-value position, each result from a macro is bound to the field-name
-independently, leading to the field being repeated or even absent:
+Tagged placeholders can provide default values that are used when no argument is provided:
 
 ```ion
-{ name: (:values) }          ⇒ { }
-{ name: (:values v) }        ⇒ { name: v }
-{ name: (:values v ann::w) } ⇒ { name: v, name: ann::w }
+(temperature {
+  degrees: (:?),
+  scale: (:? K)  // Default value is K
+})
 ```
 
-An E-expression can even be used in place of a key-value pair, in which case it must return
-structs, which are merged into the surrounding container:
+When invoking this macro, you can omit the second argument by passing `(:)` (which means "no argument"):
 
 ```ion
-{ a:1, (:values), z:3 }             ⇒ { a:1, z:3 }
-{ a:1, (:values {}), z:3 }          ⇒ { a:1, z:3 }
-{ a:1, (:values {b:2}), z:3 }       ⇒ { a:1, b:2, z:3 }
-{ a:1, (:values {b:2} {z:3}), z:3 } ⇒ { a:1, b:2, z:3, z:3 }
-
-{ a:1, (:values key "value") } ⇒ // Error: struct expected for splicing into struct
+(:temperature 96 F)   ⇒ {degrees: 96, scale: F}
+(:temperature 283 (:)) ⇒ {degrees: 283, scale: K}
 ```
 
+Note that when no value is provided for the `scale` placeholder, it uses the default value `K`.
 
-#### Splicing in template expressions
+## E-expressions in Templates
 
-The preceding examples demonstrate splicing of E-expressions into encoded data, but similar
-stream-splicing occurs within the template language, making it trivial to convert a stream to a
-list:
+Templates can contain E-expressions, which are expanded when the macro is defined, not when it's invoked.
+This allows you to use previously defined macros to construct your template.
 
 ```ion
-(macro list_of (vals*) [ (%vals) ])
-(macro clumsy_bag (elts*) { '': (%elts) })
-```
-```ion
-(:list_of)   ⇒ []
-(:clumsy_bag) ⇒ {}
-
-(:list_of 1 2 3)    ⇒ [1, 2, 3]
-(:clumsy_bag true 2) ⇒ {'':true, '':2}
-```
-
-## Mapping templates over streams: `for`
-
-Another way to produce a stream is via a mapping form. The `for` [special form](special_forms.md) evaluates a
-template once for each value provided by a stream or streams. Each time, a local variable is
-created and bound to the next value on the stream.
-
-```ion
-(macro price (a c) { amount: (%a), currency: (%c) })
-
-(macro prices (currency amounts*)
-  (.for
-    // Binding pairs
-    [(amt (%amounts))]
-    //└┬┘ └────┬───┘
-    // │       └─── stream to map over
-    // └─────────── variable name
-
-    // Template
-    (.price (%amt) (%currency))
-  )
+(:$set_macros
+  (prefix_suffix [(:?), (:?)])
+  (website_url (:prefix_suffix "https://www.amazon.com/" (:?)))
 )
```

Comment:
The `prefix_suffix` macro is not available in the macro table until after the directive closes, so it would need to be:
```suggestion
(:$set_macros (prefix_suffix [(:?), (:?)]))
(:$add_macros (website_url (:prefix_suffix "https://www.amazon.com/" (:?))))
```